zcash_primitives/src/sapling/value/sums.rs

```diff
@@ -0,0 +1,220 @@
+    }
+}
+
+impl<'a> Sum<&'a NoteValue> for Result<ValueSum, OverflowError> {
+    fn sum<I: Iterator<Item = &'a NoteValue>>(iter: I) -> Self {
+        iter.fold(Ok(ValueSum(0)), |acc, v| (acc? + *v).ok_or(OverflowError))
+    }
+}
+
+impl Sum<NoteValue> for Result<ValueSum, OverflowError> {
+    fn sum<I: Iterator<Item = NoteValue>>(iter: I) -> Self {
+        iter.fold(Ok(ValueSum(0)), |acc, v| (acc? + v).ok_or(OverflowError))
+    }
+}
+
+impl TryFrom<ValueSum> for i64 {
+    type Error = OverflowError;
+
+    fn try_from(v: ValueSum) -> Result<i64, Self::Error> {
+        i64::try_from(v.0).map_err(|_| OverflowError)
+    }
+}
+
+/// A sum of Sapling value commitment blinding factors.
+#[derive(Clone, Copy, Debug)]
+pub struct TrapdoorSum(jubjub::Scalar);
+
+impl TrapdoorSum {
+    /// Initializes a sum of `ValueCommitTrapdoor`s to zero.
+    pub fn zero() -> Self {
+        TrapdoorSum(jubjub::Scalar::zero())
+    }
+
+    /// Transform this trapdoor sum into the corresponding RedJubjub private key.
+    ///
+    /// This is public for access by `zcash_proofs`.
+    pub fn into_bsk(self) -> redjubjub::PrivateKey {
+        redjubjub::PrivateKey(self.0)
+    }
+}
+
+impl Add<&ValueCommitTrapdoor> for ValueCommitTrapdoor {
+    type Output = TrapdoorSum;
+
+    fn add(self, rhs: &Self) -> Self::Output {
+        TrapdoorSum(self.0 + rhs.0)
+    }
+}
+
+impl Add<&ValueCommitTrapdoor> for TrapdoorSum {
+    type Output = TrapdoorSum;
+
+    fn add(self, rhs: &ValueCommitTrapdoor) -> Self::Output {
+        TrapdoorSum(self.0 + rhs.0)
+    }
+}
+
+impl AddAssign<&ValueCommitTrapdoor> for TrapdoorSum {
+    fn add_assign(&mut self, rhs: &ValueCommitTrapdoor) {
+        self.0 += rhs.0;
+    }
+}
+
+impl Sub<&ValueCommitTrapdoor> for ValueCommitTrapdoor {
+    type Output = TrapdoorSum;
+
+    fn sub(self, rhs: &Self) -> Self::Output {
+        TrapdoorSum(self.0 - rhs.0)
+    }
+}
+
+impl SubAssign<&ValueCommitTrapdoor> for TrapdoorSum {
+    fn sub_assign(&mut self, rhs: &ValueCommitTrapdoor) {
+        self.0 -= rhs.0;
+    }
+}
+
+impl<'a> Sum<&'a ValueCommitTrapdoor> for TrapdoorSum {
+    fn sum<I: Iterator<Item = &'a ValueCommitTrapdoor>>(iter: I) -> Self {
+        iter.fold(TrapdoorSum::zero(), |acc, cv| acc + cv)
+    }
+}
+
+/// A sum of Sapling value commitments.
+#[derive(Clone, Copy, Debug)]
+pub struct CommitmentSum(jubjub::ExtendedPoint);
+
+impl CommitmentSum {
+    /// Initializes a sum of `ValueCommitment`s to zero.
+    pub fn zero() -> Self {
+        CommitmentSum(jubjub::ExtendedPoint::identity())
+    }
+
+    /// Transform this value commitment sum into the corresponding RedJubjub public key.
+    ///
+    /// This is public for access by `zcash_proofs`.
+    pub fn into_bvk<V: Into<i64>>(self, value_balance: V) -> Option<redjubjub::PublicKey> {
+        let value: i64 = value_balance.into();
+
+        // Compute the absolute value (failing if -i64::MAX is the value).
```

Comment:
```suggestion
        // Compute the absolute value (failing if i64::MIN is the value).
```
`-i64::MAX` is wrong. Independently of that: i64::MIN cannot occur in Zcash but could occur in the abstract protocol omitting the block size limit, and it might be able to occur in some future version of Zcash with ZSAs. It is possible to just compute bvk correctly in this case; should we do that?